dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/BomTest.java

```diff
@@ -72,4 +83,64 @@ private static void assertReachable(String url) throws IOException {
       Assert.fail("Could not reach " + url + "\n" + ex.getMessage());
     }
   }
+
+  /**
+   * Asserts that the members of the {@code bom} satisfy the no-downgrade rule. This rule means that
+   * the members have the highest versions among the dependencies of them. If there's a violation,
+   * users of the BOM would see our BOM downgrading certain dependencies. Downgrading a dependency
+   * is bad practice in general because newer versions have more features (classes and methods).
+   *
+   * <p>For example, if google-http-client 1.40.1 is in the BOM, then no other libraries in the BOM
+   * must not depend on the higher version of the google-http-client.
+   *
+   * @param bom the BOM to assert with this no-downgrade rule.
+   */
+  private static void assertNoDowngradeRule(Bom bom) throws InvalidVersionSpecificationException {
+
+    boolean violationFound = false;
+    StringBuilder errorMessages = new StringBuilder();
+    Map<String, Artifact> versionlessCoordinatesToArtifact = new HashMap<>();
+    for (Artifact artifact : bom.getManagedDependencies()) {
+      versionlessCoordinatesToArtifact.put(Artifacts.makeKey(artifact), artifact);
+    }
+    VersionScheme versionScheme = new GenericVersionScheme();
+
+    for (Artifact artifact : bom.getManagedDependencies()) {
+      ClassPathBuilder classPathBuilder = new ClassPathBuilder();
+      ClassPathResult result =
+          classPathBuilder.resolve(ImmutableList.of(artifact), false, DependencyMediation.MAVEN);
+
+      for (ClassPathEntry entry : result.getClassPath()) {
+        Artifact transitiveDependency = entry.getArtifact();
+        String dependencyVersionlessCoordinates = Artifacts.makeKey(transitiveDependency);
+        Artifact bomArtifact =
+            versionlessCoordinatesToArtifact.get(dependencyVersionlessCoordinates);
+        if (bomArtifact == null) {
+          // transitiveDependency is not part of the BOM
+          continue;
+        }
+
+        Version versionInBom = versionScheme.parseVersion(bomArtifact.getVersion());
+        Version versionInTransitiveDependency =
+            versionScheme.parseVersion(transitiveDependency.getVersion());
+
+        if (versionInTransitiveDependency.compareTo(versionInBom) <= 0) {
+          // When versionInTransitiveDependency is less than or equal to versionInBom, it satisfies
+          // the no-downgrade rule.
+          continue;
+        }
+        // No downgrade rule is violated
+        violationFound = true;
+        errorMessages.append(
+            artifact
+                + " has a transitive dependency "
+                + transitiveDependency
+                + ". This is higher version than "
+                + bomArtifact
+                + " in the BOM\n");
+      }
+    }
+
+    Assert.assertFalse(errorMessages.toString(), violationFound);
```

Comment:
Instead of having a `violationFound` variable, you can just assertFail that `!errorMessages.isEmpty()`.

Reply:
Nice idea. Updated.